funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/task_queue_subscriber.py

```diff
@@ -0,0 +1,402 @@
+import logging
+import multiprocessing
+from typing import Callable
+
+import pika
+
+logger = logging.getLogger(__name__)
+
+
+class TaskQueueSubscriber(multiprocessing.Process):
+    """The TaskQueueSubscriber is a direct rabbitMQ pipe subscriber that uses
+    the SelectConnection adaptor to enable performance consumption of messages
+    from the service
+
+    """
+
+    def __init__(
+        self,
+        pika_conn_params: pika.connection.Parameters,
+        external_queue: multiprocessing.Queue,
+        kill_event: multiprocessing.Event,
+        endpoint_uuid: str,
+        _on_message_test_hook: Callable = None,
+        exchange: str = "tasks",
+    ):
+        """
+
+        Parameters
+        ----------
+        pika_conn_params: Connection Params
+
+        external_queue: multiprocessing.Queue
+             Each incoming message will be pushed to the queue.
+             Please note that upon pushing a message into this queue, it will be
+             marked as delivered.
+
+        kill_event: multiprocessing.Event
+             This event is used to communicate a failure on the subscriber
+
+        _on_message_test_hook: Callable
+            This is a test_hook used only for testing. This is invoked on
+            every message
+
+        endpoint_uuid: endpoint uuid string
+
+        exchange: str
+            Exchange name
+        """
+
+        super().__init__()
+        self.endpoint_uuid = endpoint_uuid
+        self.conn_params = pika_conn_params
+        self.external_queue = external_queue
+        self.kill_event: multiprocessing.Event = kill_event
+        self.cleanup_complete = multiprocessing.Event()
+        self.external_callback = _on_message_test_hook
+
+        self.queue_name = f"{self.endpoint_uuid}.tasks"
+        self.routing_key = f"{self.endpoint_uuid}.tasks"
+        self.params = pika_conn_params
+        self.exchange = exchange
+        self.exchange_type = "direct"
+        self._closing = False
+        self._closed = False
```

Comment:
Is `_closed` ever used? I couldn't find a usage site.

Reply:
Removed.